baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+==============================================================================
+@File    :   SolverHist.py
+@Date    :   2020/11/30
+@Author  :   Alasdair Christison Gray
+@Description :
+"""
+
+# ==============================================================================
+# Standard Python modules
+# ==============================================================================
+from warnings import warn
+import copy
+import os
+
+# ==============================================================================
+# External Python modules
+# ==============================================================================
+import numpy as np
+
+# ==============================================================================
+# Extension modules
+# ==============================================================================
+
+
+class SolverHist(object):
+    """
+    The SolverHist class stores various useful values during a nonlinear solution in a nested dictionary structure, by
+    default, the history only stores the values from the latest iteration which are used for printing to the terminal.
+    Storing of the full solution history is enabled using the 'writeConvergenceHist' option
+    """
+
+    def __init__(self, options):
+        """
+        Create an empty history dictionary based on the current solver options
+        """
+
+        if options["writeconvergencehist"]:
+            length = int(options["maxiter"] + 1)
+            self.storeValues = True
+        else:
+            length = 1
+            self.storeValues = False
+
+        # Make a local copy of the options
+        self.options = options
+
+        # Define some defaults
+        self.defaults = {"values": {}, "valueFormat": {}}
+        self.defaults["valueFormat"]["float"] = "{:17.11e}"
+        self.defaults["values"]["float"] = np.full(length, np.nan)
+
+        self.defaults["valueFormat"]["int"] = "{:03d}"
+        self.defaults["values"]["int"] = [np.nan] * length
+
+        self.defaults["valueFormat"]["other"] = "{:^40}"
+        self.defaults["values"]["other"] = [None] * length
+
+        self.hist = {"values": {}, "valueFormat": {}, "nameFormat": {}, "varType": {}}
+        self.varsToPrint = []
+        self.latestIter = 0
+
+    def addField(self, Name, Type, ValueFormat=None, NameFormat=None, Print=False):
+        """Add a field to the solution history"""
+
+        if Print:
+            self.varsToPrint.append(Name)
+
+        if Type in [float, np.float64, np.float32]:
+            varType = "float"
+            testValue = 0.0
+        elif Type is int:
+            varType = "int"
+            testValue = 0
+        else:
+            varType = "other"
+            testValue = " "
+
+        self.hist["varType"][Name] = varType
+
+        if ValueFormat is None:
+            ValueFormat = copy.deepcopy(self.defaults["valueFormat"][varType])
+
+        if NameFormat is None:
+            testString = ValueFormat.format(testValue)
```

Comment:
Maybe this can be made more compact, you are already prescribing / entering the format as input. That could be a "hard limit" and the names might be truncated, as I assume we don't want long var names anyway. I assume here there might be some personal taste involved :upside_down_face: